### ISSUE TITLE
Better fallback defaults to support SQLCipher v3 databases

### ISSUE THREAD

**Issue by justinclift** [sqlcipher/sqlcipher#296]
How feasible would it be to have SQLCipher (eg 4.0.1) transparently support databases created by the previous release?

The current release of SQLCipher (4.0.0) needs users (or their software) to manually run a series of commands in order to open existing (created by v3) databases:

* https://www.zetetic.net/blog/2018/11/30/sqlcipher-400-release/#compatibility
* https://discuss.zetetic.net/t/upgrading-to-sqlcipher-4/3283

While we ([DB Browser for SQLite](https://github.com/sqlitebrowser/sqlitebrowser)) could add some fallback behaviour to our database opening logic, so that if a failure occurs we then try opening with the previous versions defaults... every existing application that uses SQLCipher also needs to implement this too.

It seems like the better place to have this fallback, would be in SQLCipher itself so that things "just work" and every application (or user) doesn't need to reimplement it themselves.

Potential order of operations:

1. SQLCipher library attempts to open the given (already existing) database file, using SQLCipher v4 defaults + user supplied password
2. If that succeeds, then all good, move on to other stuff
3. However, if it fails then SQLCipher should retry opening the database using the SQLCipher v3 (3.4.2?) defaults + user supplied password + any non default settings (eg different page size) supplied by the user/application
   * Note, this should not be done using the migration pragma.  The on disk format shouldn't change without explicit instruction to do so from the user, as they may be operating in an environment where other people or software still uses SQLCipher v3.x for whatever reason.
4. If that succeeds, then all good, move on to other stuff
5. If this fails, that's when SQLCipher should throw an error

Related info (tracking a macOS Homebrew issue regarding this too):

&nbsp; &nbsp; https://github.com/Homebrew/homebrew-core/issues/34991

@aguynamedben Your input here too would be useful. :smile:

**Comment by justinclift** [comment]
> I can see how that could create problems if you're just installing 'sqlcipher' from Homebrew and don't care what the version is or if it changed.

Understood.  From a developer perspective, it's sometimes difficult to see things the way end users experience them.

This seems to be one of those cases, where from a developer perspective this is all "as per normal".  As you're mentioning though, there are indeed end user use cases where it's (from their perspective) now "broken". :frowning:

**Comment by brodybits** [comment]
> there are indeed end user use cases where it's (from their perspective) now "broken". 😦

That is the unfortunate result of "breaking changes"; I do not recall seeing much notice or beta testing period (@sqlcipher team please correct me if I am mistaken here).

I do think end-user applications like Homebrew *should* be more careful about breaking changes like this one. At least a warning with a workaround would have been nice.

What if the SQLCipher CLI tool would try fallback to v3 by default, at least for a limited transition period?

**Comment by sjlombardo** [comment]
Hello @justinclift - thanks for taking the time to prepare this thoughtful suggestion. I'd like to address it from two different angles (technical / non-technical) and then propose some options.

Unfortunately, from a technical perspective, it would be difficult to make this proposal work. In the current design the setup of SQLCipher (i.e. preparing the context that configures SQLCipher) occurs separately from actually deriving keys and decrypting data. Put simply, when you call `PRAGMA key` or `sqlite3_key` the library doesn't really touch the database. It just prepares an internal context based on certain defaults, at which point the various settings can modified by the SQLCipher-specific configuration PRAGMAs. It is not until a statement executes that would read the database that SQLCipher is invoked to prepare cryptographic primitives, derive keys, decrypt data, etc. 

By that time in the lifecycle SQLCipher no longer has the handle to the database (and internals) and everything up-stack in SQLite has already been setup. If the attempt fails, it’s not simple to reset things cleanly or reattach a new context. Plus, once the decryption fails, the SQLite database handle is already inoperable. It is cleaner, both for the SQLite API and SQLCipher, for the calling application to close the connection and retry. 

Finally, there could be a significant performance impact here. Even if we could easily retry, SQLCipher would have no way of knowing whether the reason decryption failed was due to invalid key material or a non-default setting. Thus, each time an incorrect key was provided, SQLCipher would need to exhaustively try settings for 4, 3, 2, and 1 before determining that the key is actually incorrect. Even in the case of a correct key with an older version performance would be reduced by requiring multiple tries to select the right version.

Of course, changing something like this isn't strictly impossible. For instance, it might be possible to someday restructure the SQLCipher codebase so that it probes the database when the key is set rather than on first use. However, a change like that would be relatively non-trivial to implement, could potentially impact many test cases, and would be considered a higher risk change. As such it wouldn't be a good fit for a patch release of SQLCipher 4.0. 

From a non-technical perspective, there is the question of whether this would be a worthwhile change for most users. In a perfect world, transparent backwards compatibility would be a nice feature, but it is not something that SQLCipher has ever offered. In SQLCipher’s 10 years, we have introduced breaking changes with every major release, starting from version 0.x to the stable 1.0 release through this latest version 4.0. We have always committed to compatibility solely within the major version for SQLCipher. We always communicate that major releases contain breaking changes so that applications can be aware of it. Finally, we have always implemented ways for applications to cope with the upgrade, either via explicitly enabling backwards compatibility or transparently migrating data. 

This approach has worked well in the past and it has a few benefits:

1. It strongly encourages migration to newer and more secure database settings
2. It makes developers and users cognizant of their implementation choices if they decide not to upgrade (i.e. they need to explicitly opt-in to being less secure via PRAGMA)
3. It provides a well-established framework for us, as developers, to group substantial changes into specific releases

With respect to specific compatibility scenarios in an application, the approach @billymeltdown outlined makes a lot of sense. If an application upgrades to SQLCipher 4 and wants to support both 4 and 3 at the same time, we would recommend it try with default settings and then, if that fails, try to open with 3 settings. When it can be opened with 3 settings, perhaps it's even a good opportunity to ask the user if they'd like to migrate to the new version automatically for improved security? This way the application controls which versions of SQLCipher databases it wants to support. It chooses how it will handle the tradeoff of compatibility vs. migration.

If an application is using a shared system library for SQLCipher (so it doesn't control the SQLCipher version directly) it can still query SQLCipher to find out what version is in use via `PRAGMA cipher_version`. This way it can make an intelligent decision about how to implement backwards compatibility (or even better offer the option to migrate).

For shell users (per @brodybits comments), while it can be annoying to repetitively re-enter compatibility statements, we do offer the cipher_default_* pragma variants. If a shell user always wants to operate in 3.x compatibility mode, they can tuck those pragmas in `~/.sqliterc` and the shell will automatically load it at the start of the session. It would then operate no differently than the original SQLCipher 3 shell.

With respect to homebrew specifically, perhaps this would be a good place to use the `caveats` feature to communicate compatibility changes to users who upgrade their SQLCipher install? 

That said, we will admit that, especially with the current 4.0 release, it is somewhat cumbersome to configure backwards compatibility because of the sheer number of settings that have changed. Thus, as a compromise here, I would like to propose that we add two new PRAGMAs that greatly simplify backwards compatibility settings: 

`PRAGMA cipher_compatibility = VERSION` - By calling this single pragma and passing in 1, 2, or 3, SQLCipher will operate with the default settings consistent with that version for the current connection

`PRAGMA cipher_default_compatibility = VERSION` - By calling this single pragma and passing in 1, 2, or 3, SQLCipher will operate with the default settings consistent for all connections opened after the pragma is set.

Our hope would be that these could greatly simplify compatibility settings for an application. A developer would no longer need to know any details of algorithms, page sizes, or KDF functions. Instead, they'd only need to call the given PRAGMA with a numeric major version number. An application could simply cycle/loop through the versions they support for compatibility settings, without even knowing what the defaults were for each version. Likewise for shell users a single default command in `.sqliterc` or shell history would enable backwards compatibility as well. We can easily extend this approach for future versions, too. 

To summarize, I don't think that transparent backwards compatibility is something we could support in the short term (though we can certainly continue to weigh the pros and cons and consider it for the future). In the meantime though, I do think we can simplify compatibility via PRAGMA in a point release. Let us know what you think.

**Comment by aguynamedben** [comment]
I haven't worked on the upgrade for app yet, but I just wanted to thank you for working hard on SQLCipher and providing a thoughtful response. We're shipping an Electron app and I think it will just be a matter of having the new version of our app update the SQLCipher database version with the appropriate commands (i.e. full out, forced v4 upgrade). The command line/homebrew issue is a nuisance but not a show stopper.

I think it's reasonable given my modest experience with Postgres/MySQL/SQLite on-disk formats to expect that a new version (i.e. SQLCipher 4) not be backwards compatible with on-disk data from a previous version (i.e. SQLCipher 3). I think the area to optimize is making auto-updating the on-disk data format very easy. I.e. if SQLCipher 4 tries to open a SQLCipher 3 database, maybe it can auto-upgrade it if the right setting is turned on. Or hopefully it's just 1-2 simple commands to migrate the data from the old version to the new version. This provides faux backward compatibility by upgrading.

I'll give more feedback in a few weeks when I actually implement the changes in our app, but thanks for the quality response.

**Comment by aguynamedben** [comment]
@justinclift I see and feel for why this is a challenging issue for DBBrowser for SQLite (which I use and love). You want to support multiple versions of SQLCipher dynamically. Our app is more in the "normal" app category where we can just auto-upgrade the end user. The user isn't aware of the database in our use case.

**Comment by justinclift** [comment]
> You want to support multiple versions of SQLCipher dynamically.

Kind of.  With DB Browser for SQLite (DB4S), it seems like many of our users aren't all that deeply IT literate.  At least when they're in the "new user" stage. :wink:

From what I can tell Homebrew, also has lots of not-real-deep-with-IT users as well.  My concern is more with Homebrew having (all in good faith) automatically bumped the SQLCipher version.

So users that were previously receiving SQLCipher 3 are now receiving SQLCipher 4, making their already saved databases need additional manual steps to continue er... working.

(After discovering this automatic version bump yesterday :wink:) I was kind of expecting that it have had a bad result.  But looking through the Homebrew issue tracker there doesn't seem to be anyone reporting problems with SQLCipher:

* https://github.com/Homebrew/homebrew-core/issues?utf8=%E2%9C%93&q=is%3Aissue+sqlcipher

So maybe I'm concerned about a non-issue then?  I guess the Homebrew users who do use SQLCipher have been figuring it out already. :wink:

With DB4S, (as of last night) our nightly builds [have initial support](https://github.com/sqlitebrowser/sqlitebrowser/issues/1640#issuecomment-446578193) for both SQLCipher 3 & 4:

![screenshot_20181212_135049](https://user-images.githubusercontent.com/3153504/49870940-25fe6f00-fe15-11e8-869b-44a48233114c.png)

Maybe I'm over-thinking things and we should just drop this?

Although the potential new compatibility PRAGMA's suggested in @sjlombardo's excellent post above do sound useful for simplifying things... if I'm just over-thinking stuff here and there isn't really any real world user impact, then implementing them would just be a waste of Zetetic's time.  Not sure. :confused:

**Comment by sjlombardo** [comment]
Hi @justinclift - that application interface looks great, it should be very clear to users and will allow a great deal of control for custom cases too. 

We just wrapped up some changes to the compatibility PRAGMAs I mentioned. They will be a solid addition for the next patch release and will make things easier and more accessible for people across the board.

I don't feel like this is a non-issue or that you're over-thinking things; these are good suggestions and I'm sure that we will have some growing pains with the move to SQLCipher 4. I'm just hopeful that we can address them sufficiently with good communication, support, and coordination. We'll definitely keep this in mind for the future.

**Comment by justinclift** [comment]
No worries. :smile:

That dialog was @MKleusberg's good idea. :smile:

Am thinking we should probably add a way for users to migrate their on disk format from v3 to v4.

Maybe - when DB4S is built with SQLCipher 4 - we should display a checkbox under the "Encryption Settings" drop downs for migrating?

* [ ] Migrate database to SQLCipher 4?

It'd probably also need to be greyed out if the user hasn't indicated they're "coming from" SQLCipher 3.  That "coming from" SQLCipher 3 bit... suspecting there might be gotchas there, especially if people aren't using the default SQLCipher 3 settings.  Still, this would probably be a decent start.

@MKleusberg What do you reckon? :wink:

**Comment by mtissington** [comment]
@sjlombardo The additional PRAGMAs will be a great help - thanks.

**Comment by sjlombardo** [comment]
Hi @justinclift - [SQLCipher 4.0.1](https://www.zetetic.net/blog/2018/12/18/sqlcipher-401-release/) includes the new PRAGMAs (cipher_compatibility and cipher_default_compatibility) we talked about here.
